library/psa_crypto_slot_management.h

```diff
@@ -136,9 +138,35 @@ static inline psa_status_t psa_lock_key_slot(psa_key_slot_t *slot)
     return PSA_SUCCESS;
 }
 
+/** Lock a key slot for writing.
+ *
+ * This function sets the slot's lock_count to SIZE_MAX.
+ * If multi-threading is enabled, the caller must hold the
+ * global key slot mutex.
+ *
+ * \param[in] slot  The key slot.
+ *
+ * \retval #PSA_SUCCESS
+               The key slot lock counter was set to SIZE_MAX.
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
+ *             The lock counter was already in a writing state.
+ */
+static inline psa_status_t psa_lock_key_slot_for_writing(psa_key_slot_t *slot)
+{
+    if (slot->lock_count == SIZE_MAX) {
+        return PSA_ERROR_CORRUPTION_DETECTED;
+    }
+    slot->lock_count = SIZE_MAX;
+
+    return PSA_SUCCESS;
+}
+
 /** Unlock a key slot.
  *
  * This function decrements the key slot lock counter by one.
+ * If slot is in a WRITING state, this will fail.
```

Comment:
Having "unlock key slot" fail when a write lock is held against the slot makes sense while reading this PR, as the "default" lock type is clearly "read lock", but when we come back to it in six months time, I can see this leading to confusion.

Should we rename this at this point (not later), so that it's always explicit whether we're taking a read or write lock?  `psa_acquire_key_slot_read_lock()`,  `psa_release_key_slot_read_lock()`, `psa_acquire_key_slot_write_lock()` and  `psa_release_key_slot_write_lock()`?

Reply:
This is something that I wanted to do but wasn't sure if it was allowed. If it is, then I definitely think we should rename.

Reply:
As they're in a header in `library/` rather than under `include/` they are private functions, so we are free to change that part of the API